modules/images/image-loading-optimization/storage/data.php

```diff
@@ -0,0 +1,311 @@
+	 * @param int $ttl Expiration TTL in seconds.
+	 */
+	return (int) apply_filters( 'ilo_page_metric_freshness_ttl', DAY_IN_SECONDS );
+}
+
+/**
+ * Determines whether the current response can be optimized.
+ *
+ * Only search results are not eligible by default for optimization. This is because there is no predictability in
+ * whether posts in the loop will have featured images assigned or not. If a theme template for search results doesn't
+ * even show featured images, then this isn't an issue.
+ *
+ * @return bool Whether response can be optimized.
+ */
+function ilo_can_optimize_response(): bool {
+	$able = ! is_search();
+
+	/**
+	 * Filters whether the current response can be optimized.
+	 *
+	 * @param bool $able Whether response can be optimized.
+	 */
+	return (bool) apply_filters( 'ilo_can_optimize_response', $able );
+}
+
+/**
+ * Gets the normalized query vars for the current request.
+ *
+ * This is used as a cache key for stored page metrics.
+ *
+ * TODO: For non-singular requests, consider adding the post IDs from The Loop to ensure publishing a new post will invalidate the cache.
+ *
+ * @return array Normalized query vars.
+ */
+function ilo_get_normalized_query_vars(): array {
+	global $wp;
+
+	// Note that the order of this array is naturally normalized since it is
+	// assembled by iterating over public_query_vars.
+	$normalized_query_vars = $wp->query_vars;
+
+	// Normalize unbounded query vars.
+	if ( is_404() ) {
+		$normalized_query_vars = array(
+			'error' => 404,
+		);
+	}
+
+	return $normalized_query_vars;
+}
+
+/**
+ * Gets slug for page metrics.
+ *
+ * @see ilo_get_normalized_query_vars()
+ *
+ * @param array $query_vars Normalized query vars.
+ * @return string Slug.
+ */
+function ilo_get_page_metrics_slug( array $query_vars ): string {
+	return md5( wp_json_encode( $query_vars ) );
+}
+
+/**
+ * Compute nonce for storing page metrics for a specific slug.
+ *
+ * This is used in the REST API to authenticate the storage of new page metrics from a given URL.
+ *
+ * @see wp_create_nonce()
+ * @see ilo_verify_page_metrics_storage_nonce()
+ *
+ * @param string $slug Page metrics slug.
+ * @return string Nonce.
+ */
+function ilo_get_page_metrics_storage_nonce( string $slug ): string {
+	return wp_create_nonce( "store_page_metrics:{$slug}" );
+}
+
+/**
+ * Verify nonce for storing page metrics for a specific slug.
+ *
+ * @see wp_verify_nonce()
+ * @see ilo_get_page_metrics_storage_nonce()
+ *
+ * @param string $nonce Page metrics storage nonce.
+ * @param string $slug  Page metrics slug.
+ * @return int 1 if the nonce is valid and generated between 0-12 hours ago,
+ *             2 if the nonce is valid and generated between 12-24 hours ago.
+ *             0 if the nonce is invalid.
+ */
+function ilo_verify_page_metrics_storage_nonce( string $nonce, string $slug ): int {
+	return (int) wp_verify_nonce( $nonce, "store_page_metrics:{$slug}" );
+}
+
+/**
+ * Unshift a new page metric onto an array of page metrics.
```

Comment:
```suggestion
 * Unshifts a new page metric onto an array of page metrics.
```

There may be some more functions not using third-person singular, would be good to double check.

Reply:
Fixed in eadec5592a8b103e3522fa9f29b75e5b8dcea2e9 and 02dbb359fa726ed889475f9cad66bb362f9b36e8